rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+    fn tokenize(mut self, text: &str) -> Vec<String> {
+        let mut stream = self.token_stream(text);
+        let mut tokens = Vec::new();
+        while stream.advance() {
+            let token = stream.token();
+            tokens.push(token.text.clone());
+        }
+        tokens
+    }
+}
+
+/// Basic stats about an ngram index
+#[derive(Serialize)]
+struct NGramStatistics {
+    num_ngrams: usize,
+}
+
+/// The row ids that contain a given ngram
+#[derive(Debug)]
+struct NGramPostingList {
+    bitmap: RoaringTreemap,
+}
+
+impl DeepSizeOf for NGramPostingList {
+    fn deep_size_of_children(&self, _: &mut deepsize::Context) -> usize {
+        self.bitmap.serialized_size() as usize
+    }
+}
+
+impl NGramPostingList {
+    fn try_from_batch(batch: RecordBatch) -> Result<Self> {
+        let bitmap_bytes = batch.column(0).as_binary::<i32>().value(0);
+        let bitmap =
+            RoaringTreemap::deserialize_from(bitmap_bytes).map_err(|e| Error::Internal {
+                message: format!("Error deserializing ngram list: {}", e),
+                location: location!(),
+            })?;
+        Ok(Self { bitmap })
+    }
+
+    fn intersect<'a>(lists: impl IntoIterator<Item = &'a Self>) -> RoaringTreemap {
+        let mut iter = lists.into_iter();
+        let mut result = iter
+            .next()
+            .map(|list| list.bitmap.clone())
+            .unwrap_or_default();
+        for list in iter {
+            result &= &list.bitmap;
+        }
+        result
+    }
+}
+
+/// Reads on-demand ngram posting lists from storage (and stores them in a cache)
+struct NGramPostingListReader {
+    reader: Arc<dyn IndexReader>,
+    cache: Cache<u32, Arc<NGramPostingList>>,
+}
+
+impl DeepSizeOf for NGramPostingListReader {
+    fn deep_size_of_children(&self, _: &mut deepsize::Context) -> usize {
+        self.cache.weighted_size() as usize
+    }
+}
+
+impl std::fmt::Debug for NGramPostingListReader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NGramListReader")
+            .field("cache_entry_count", &self.cache.entry_count())
+            .finish()
+    }
+}
+
+impl NGramPostingListReader {
+    #[instrument(level = "debug", skip(self))]
+    pub async fn ngram_list(&self, token_id: u32) -> Result<Arc<NGramPostingList>> {
+        Ok(self
+            .cache
+            .try_get_with(token_id, async move {
+                let batch = self
+                    .reader
+                    .read_range(
+                        token_id as usize..token_id as usize + 1,
+                        Some(&[POSTING_LIST_COL]),
+                    )
+                    .await?;
+                Result::Ok(Arc::new(NGramPostingList::try_from_batch(batch)?))
+            })
+            .await
+            .map_err(|e| Error::io(e.to_string(), location!()))?)
+    }
+}
+
+/// An ngram index
+///
+/// This index stores a mapping from ngrams to the row ids that contain them.
```

Comment:
I think it might be useful to mention that this is implemented as an inverted index, even though it's not the same as an index_type=INVERTED in our code base.

Reply:
Revised the comment